lib/rpush/daemon/apns/delivery.rb

```diff
@@ -26,8 +26,10 @@ def perform
 
         def batch_to_binary
           payload = ""
-          @batch.each_notification do |notification|
-            payload << notification.to_binary
+          @batch
```

Comment:
This will break APNs error handling functionality in [Rpush::Daemon::Dispatcher::ApnsTcp][1].

According to [The Binary Interface and Notification Format][2] document, "If you send a notification that is malformed or otherwise unintelligible, APNs returns an error-response packet and closes the connection. Any notifications that you sent after the malformed notification using the same connection are discarded, and must be resent. [...] The packet has [...] the notification identifier of the malformed notification".

That "notification identifier" will be equal to `frame_id` in [Rpush::Client::ActiveModel::Apns::Notification#to_binary][3]. Currently this ID equals to a `Rpush::Apns::Notification#id`, which helps `ApnsTcp` to determine the failed notification and mark it accordingly. But you've used an index as a frame_id here, so `ApnsTcp` won't be able to fail notifications correctly.

This is how I see the possible fix:

1. Encode both Notification.id and device_token into a frame_id
2. In `ApnsTcp`:
  - Parse notification_id returned by APNs to get the notification id and an index of the failed device_token
  - Find corresponding notification
  - Take all device_tokens after the failed one and create a new `Rpush::Apns::Notification` from them
  - Mark the old notification as failed

And you should also ensure that all the notifications sent after the failed in the same batch are marked as retryable.

Or maybe since a single notification can now contain multiple device tokens it is possible to switch off batch_deliveries for `Rpush::Daemon::Apns`?

[1]: https://github.com/rpush/rpush/blob/master/lib/rpush/daemon/dispatcher/apns_tcp.rb#L119
[2]: https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPhttps://github.com/rpush/rpush/blob/master/lib/rpush/client/active_model/apns/notification.rb#L77G/BinaryProviderAPI.html#//apple_ref/doc/uid/TP40008194-CH13-SW4
[3]: https://github.com/rpush/rpush/blob/master/lib/rpush/client/active_model/apns/notification.rb#L77

Reply:
I thought that `Apns::Feedback` was the only error handling `Apns` had.
`Rpush::Daemon::Dispatcher::ApnsTcp#deliver_buffer` can help map `frame_id` to `notification` as it is used in `#handle_error` anyway.